src/sinks/datadog/traces/stats.rs

```diff
@@ -0,0 +1,532 @@
+        return 1.0;
+    }
+    let mut parent_id_to_child_weigth = BTreeMap::<i64, f64>::new();
+    let mut span_ids = Vec::<i64>::new();
+    for s in spans.iter() {
+        let parent_id = match s.get("parent_id") {
+            Some(Value::Integer(val)) => *val,
+            _ => 0,
+        };
+        let span_id = match s.get("span_id") {
+            Some(Value::Integer(val)) => *val,
+            _ => 0,
+        };
+        let sample_rate = s
+            .get("metrics")
+            .and_then(|m| m.as_object())
+            .map(|m| match m.get(SAMPLING_RATE_KEY) {
+                Some(Value::Float(v)) => {
+                    let sample_rate = v.into_inner();
+                    if sample_rate <= 0.0 || sample_rate > 1.0 {
+                        1.0
+                    } else {
+                        1.0 / sample_rate
+                    }
+                }
+                _ => 1.0,
+            })
+            .unwrap_or(1.0);
+        if parent_id == 0 {
+            return sample_rate;
+        }
+        span_ids.push(span_id);
+        parent_id_to_child_weigth.insert(parent_id, sample_rate);
+    }
+    // We remove all span that do have a parent
+    span_ids.iter().for_each(|id| {
+        parent_id_to_child_weigth.remove(id);
+    });
+    // There should be only one value remaining, the weigth from the root span
+    if parent_id_to_child_weigth.len() != 1 {
+        debug!("Didn't reliably find the root span.");
+    }
+
+    *parent_id_to_child_weigth
+        .values()
+        .next()
+        .unwrap_or_else(|| {
+            debug!("Root span was not found");
+            &1.0
+        })
+}
+
+pub(crate) fn compute_apm_stats(key: &PartitionKey, traces: &[TraceEvent]) -> StatsPayload {
+    let mut aggregator = Aggregator::new();
+    traces.iter().for_each(|t| aggregator.handle_trace(key, t));
+    StatsPayload {
+        agent_hostname: key.hostname.clone().unwrap_or_else(|| "".to_string()),
+        agent_env: key.env.clone().unwrap_or_else(|| "".to_string()),
+        stats: aggregator.get_client_stats_payload(),
+        agent_version: key.agent_version.clone().unwrap_or_else(|| "".to_string()),
+        client_computed: false,
+    }
+}
+
+// Stats struct come from .proto and field name are set to match go
+// See this https://github.com/DataDog/datadog-agent/blob/b5bed4d/pkg/trace/pb/stats_gen.go for the reference go code
+#[derive(Debug, PartialEq, Deserialize, Serialize)]
+pub(crate) struct StatsPayload {
+    #[serde(rename = "AgentHostname")]
+    pub(crate) agent_hostname: String,
+    #[serde(rename = "AgentEnv")]
+    pub(crate) agent_env: String,
+    #[serde(rename = "Stats")]
+    pub(crate) stats: Vec<ClientStatsPayload>,
+    #[serde(rename = "AgentVersion")]
+    pub(crate) agent_version: String,
+    #[serde(rename = "ClientComputed")]
+    pub(crate) client_computed: bool,
+}
+
+#[derive(Debug, PartialEq, Deserialize, Serialize)]
+pub(crate) struct ClientStatsPayload {
```

Comment:
Same here:

```suggestion
#[derive(Debug, PartialEq, Deserialize, Serialize)]
#[serde(rename_all = "PascalCase")]
pub(crate) struct ClientStatsPayload {
```

(`runtime_id` and `container_id` still need manual renames)